images/hub/jupyterhub_config.py

```diff
@@ -177,6 +178,25 @@ def get_config(key, default=None):
     c.DummyAuthenticator.password = get_config('auth.dummy.password', None)
 elif auth_type == 'tmp':
     c.JupyterHub.authenticator_class = 'tmpauthenticator.TmpAuthenticator'
+elif auth_type == 'custom':
+    # full_class_name looks like "myauthenticator.MyAuthenticator".
+    # To create a docker image with this class availabe, you can just have the
+    # following Dockerifle:
+    #   FROM jupyterhub/k8s-hub:v0.4
+    #   RUN pip3 install myauthenticator
+    full_class_name = get_config('auth.custom.class_name')
+    c.JupyterHub.authenticator_class = full_class_name
+
+    (module, class_name) = full_class_name.split('.')
+    the_module = importlib.import_module(module)
+    klass=getattr(the_module, class_name)
+
+    class_params = get_config('auth.custom.params')
```

Comment:
One of the tricks of this is we have config systems on top of config systems on top of config systems, and it gets a little messy when we expose the guts of one to another.

> Now everything under auth.custom never reaches Kubernetes, since helm filters it out. If this part of the config was under extraConfig, I wouldn't be able to filter it, I'd have to find it and comment it out under "extraConfig".

traitlets config for classes that aren't used is ignored, so changing `type` to dummy and leaving configuration of HashAuth in extraConfig is the same as leaving it in auth.custom and not using it. It has no effect unless `authenticator_class = HashAuthenticator`.

> the benefit that I see there is that there's only one place in the config that will use this syntax of configuring a class.

I don't think it's true that it's the only place where a class could be configured (KubeSpawner, JupyterHub, and Proxy may all be configured in the same way), but I see the user benefit of having the auth config in the same place as the auth choice.

So I'm +.5 on putting the custom authenticator config under auth.custom.config (or .traits or whatever).

@yuvipanda it's probably worthwhile having extraConfig emit JSON. I'm a bit wary of switching based on type (maybe different keys for json and python), but I think writing JSON config directly from the YAML seems desirable to me, since we are already in a json-equivalent context.